### DISCUSSION TITLE
Can a client run alone without ever connecting to an API?

### DISCUSSION THREAD

**Discussion by tm6907** [cmu-sei/GHOSTS#377] [Q&A] [answered]
Wondering if a Linux client can run activity alone without ever having connected to a API. 

If so what should be written in the application.json file? Right now I have it as it was out-of-the-box, but doesn't work. Tried on Rocky9, Rocky8 and Centos8. 

I did have my Rocky8 client working yesterday but that was a client who was connected to an API in the past i.e. had an instance ID. 
I removed that instance folder and set the value in application.json back to "localhost:5000/api" as it is from a fresh download and now it doesn't work. My theory is it only worked because it had been connected in the past with that instance ID, even if you change around the connection to localhost rather than the API.

**Comment by sei-dupdyke** [comment]
Yes, clients will work without the API, you just won't be able to send them commands or collect activity history, but every activity should occur and be logged locally to ./logs/.

You can turn off the API stuff in the client via the application.json file.

When you say it doesn't work, what specifically is not working?

**Reply by tm6907** [reply]
So it attempts to connect, halts for a minute then dumps a load of errors and stops. Looks like this...
<img width="982" alt="Screenshot 2024-07-22 at 11 14 05" src="https://github.com/user-attachments/assets/0bd6730a-f237-431d-bab6-b12e6e48678a">

Tried setting the "ApiRootUrl" from localhost to the IP the client machine and to 127.0.0.1 and get the same result. 

It seems a Windows client can run standalone but not Linux, is it because its ran with dotnet on a .dll file instead of a .exe ?

**Reply by tm6907** [reply]
Nevermind, managed to fix it by setting ID IsEnabled to false

**Reply by tm6907** [reply]
Got a new issue now, its saying "out of memory" even though the system isn't. 
<img width="1199" alt="image" src="https://github.com/user-attachments/assets/a9b26c4e-a6ff-4b8e-be71-70b011471a95">
<img width="671" alt="image" src="https://github.com/user-attachments/assets/f01cdebf-1dd0-4879-9571-9be28983ffcf">
Not had this issue before now

**Reply by sei-dupdyke** [reply]
Have not seen this. I'm not familiar with Packer.

With Docker, I have definitely come across weird behavior when it starts to run out of memory. This is solved with a `docker system prune` which typically clears up copious GB of space :)

**Reply by tm6907** [reply]
Docker shouldn't come into play here since its just a client right? Packer is just the user name. 

I reverted the change I made in application.json but get the same issue as before. I changes the isEnabled value from true to false for the Id block. It did work in testing before but now doesn't.

**Reply by sei-dupdyke** [reply]
Right, docker not involved with client.

What is in system-lv_root? — that is a lot of space.

**Reply by tm6907** [reply]
I'm not sure but that shouldn't matter really if its not full. 

Turns out rebooting the VM does fix this "out of memory" issue but it does keep coming back. Still don't know whats causing it but it seems to be Ghosts specifically. 

I'll try a clean install just incase it corrupted or something.

**Reply by tm6907** [reply]
Yeah still stuck, not sure whats happening

**Comment by sei-dupdyke** [comment]
Hmm.

- What version are you running? `dotnet ghosts.client.linux.dll --version`
- Are there any errors or fatals in ./logs/application.log?
- What handlers are you running? Browser? Something else?
- Try running in debug mode: `dotnet ghosts.client.linux.dll --debug` and see if that tells us more in the logs please.

**Reply by tm6907** [reply]
Seems to work inconsistently. Rebooting does fix it temporarily. 

Version:  
![image](https://github.com/user-attachments/assets/6cf927ed-b49e-4fc9-a563-40dafa90d3eb)

Logs show nothing of note. Shows its working at the moment because I rebooted again, but it wasn't working 30 mins ago. 

The only handler is for Firefox, with the basic timeline that comes with the download. Perhaps its to with the geckodriver again? Looking at the logs at past runs it shows errors relating to that but it also has been working so idk. Usually a problem with the driver is shown in the command line instead of this "out of memory" error. 

I'll wait for it to break again and post the full logs.

**Reply by tm6907** [reply]
I'm trying to run it remotely with an Ansible script, which is how I discovered this error

**Comment by sei-dupdyke** [comment]
Ah, so we have seen this weird Firefox behavior (on Windows IIRC) where it is creating an enormous amount of temp files. Can you please check the FF directories (app and profile) to see if this is the case?

**Reply by tm6907** [reply]
Yeah didn't take long to break again 
<img width="1149" alt="image" src="https://github.com/user-attachments/assets/cad87600-711d-440a-8d1e-865ff3e59618">

Got this weirdness too, not seen this one before :D 
![image](https://github.com/user-attachments/assets/038ba114-748e-4649-abea-6c85826b7aa9)

Logs are just going on about the driver 
<img width="1500" alt="image" src="https://github.com/user-attachments/assets/4952c00c-9c96-40c8-8e14-9e2d9a78862d">


I'll take a look in the FF files now

**Reply by tm6907** [reply]
Well the ./cache/mozilla/firefox and /.mozilla/firefox are empty which is the default location for temp files I believe. I wasn't sure where else to look so I just remove firefox and related files then reinstalled it. 

Ghosts is running fine at the moment, but it might break again soon like before.

**Reply by tm6907** [reply]
Still broken, any idea which file in FF is causing it? If this is the issue after all?

**Comment by sei-dupdyke** [comment]
Please try this:

run a `top -n 1` and log all that to a log file also inject a timestamp:

```sh
top -n 1;date > mylogfile.txt
```

**Reply by tm6907** [reply, marked as answer]
I think my issue stems from my Ansible code. I've been trying to start Ghosts with a ansible task and not seeing anything I would expect because its effectively acting on a shell I can't see. So when I've been starting & restarting Ghosts over and over either by Ansible or manually, its clogging up memory leading to the 'out of memory' I've been seeing and why rebooting solves all issues. 

I tired making a script for Ansible to run that instead but no help. I think I'll have to keep my Ansible playbook to just configuring the client before starting it up.
